### ISSUE TITLE
Access violation while loading mod_spatialite

### ISSUE THREAD

**Issue by fromm1990** [aspnet/Microsoft.Data.Sqlite#471]
Hi I have an issue regarding loading the mod_spatialite extension. Here is how I try to do it:

```csharp
var connectionStringBuilder = new SqliteConnectionStringBuilder();
connectionStringBuilder.DataSource = filePath;

using (var connection = new SqliteConnection(connectionStringBuilder.ConnectionString))
{
    using (var command = connection.CreateCommand())
    {
        connection.Open();
        connection.EnableExtensions();
        command.CommandText = @"SELECT load_extension('mod_spatialite');";
        command.ExecuteNonQuery();
    }
}
```

As soon as the  `command.ExecuteNonQuery();` is executed, i get the following error

`The program '[14956] dotnet.exe' has exited with code -1073741819 (0xc0000005) 'Access violation'.`

I am using .NET Core 2.0 but i have been able to reproduce this "error" in .NET Framework 4.6.1. 
Furthermore, I use the Microsoft.Data.Sqlite 2.0.0 but is getting the same behaviour when using the 1.1.1 version.

Am I doing something wrong here, or is this a bug?

**Comment by bricelam** [comment]
I wonder if you're mixing x64 and x86 libraries. Does adding this to your `*.csproj` make it work?

```
<PropertyGroup>
  <RuntimeIdentifier>win10-x86</RuntimeIdentifier>
</PropertyGroup>
```

**Comment by fromm1990** [comment]
This seems to fix the issue when I swap out my 64 bit version of mod_spatialite with a 32 bit one. Is it not possible to use a 64 bit version of SQLite?

**Comment by bricelam** [comment]
You may have installed the 32-bit version of the .NET Core SDK (run `where dotnet` to confirm)

**Comment by fromm1990** [comment]
It seems like i got the 64-bit version of .NET Core SDK (C:\Program Files\dotnet)?

**Comment by bricelam** [comment]
Weird. Search your `*.csproj` files for `RuntimeIdentifier` or `Platform` something is causing it to be x86 instead of the default of `x64`

**Comment by fromm1990** [comment]
My `*.csproj` is rather simple

```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>netcoreapp2.0</TargetFramework>
    <RuntimeIdentifier>win10-x86</RuntimeIdentifier>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Microsoft.Data.SQLite" Version="2.0.0" />
    <PackageReference Include="Newtonsoft.Json" Version="10.0.3" />
  </ItemGroup>

</Project>
```

However, using the 64 bit version of `mod_spatialite.dll` without `<RuntimeIdentifier>win10-x86</RuntimeIdentifier>` will make dotnet.exe crash with the `Access violation` exception.

Normally, mixing up 32 bit 64 bit packages results in the following exception `Microsoft.Data.Sqlite.SqliteException: 'SQLite Error 1: '%1 is not a valid Win32 application.
'.'`

Does my `.csproj` give you a clue towards the problem?

**Comment by bricelam** [comment]
Hmm, looks like a problem specific to x64. @ericsink, is there anything special about e_sqlite3.dll that might prevent it from working with `mod_spatialite` on x64?

**Comment by bricelam** [comment]
(P.S. I have very little knowledge of how SQLite extensions work under the covers.)

**Comment by ericsink** [comment]
Not that I know of.

But a mismatched calling convention would be among my first suspicions.

Dynamic loading of extensions is not a frequently used feature.

Where did your build of mod_spatialite come from?

**Comment by fromm1990** [comment]
I got it from this url http://www.gaia-gis.it/gaia-sins/windows-bin-amd64/

**Comment by ericsink** [comment]
I'm not sure what calling convention that lib is built with.  Nor can I be certain right now that a mismatched calling convention is in fact the problem.  This is going to require more investigation.

**Comment by fromm1990** [comment]
Found a solution to the problem. Using the 64 bit mod_spatialite package distributed at http://www.gaia-gis.it/gaia-sins/windows-bin-amd64/ with sqlite.exe, does not seem to cause any problem. However, when using Microsoft.Data.Sqlite the libstdc++_64-6.dll dependency starts to cause the `Access violation` issue. 

By following this guide http://blog.jrg.com.br/2016/04/25/Fixing-spatialite-loading-problem/ i manged to swap the `libstdc++_64-6.dll` to a version not triggering this issue. I found `libstdc++-6.dll` shipped in `MinGW-W64 GCC-5.4.0 -x86_64-win32-seh`, which can be found [here](https://sourceforge.net/projects/mingw-w64/files/Toolchains%20targetting%20Win64/Personal%20Builds/mingw-builds/5.3.0/threads-posix/seh/), to solve my issue so far.

**OBS**
Remember to change the name from `libstdc++-6.dll` to `libstdc++_64-6.dll` and include its `libgcc_s_seh-1.dll` dependency  (without changing the name of the latter).
